arch/x86/core/ia32/crt0.S

```diff
@@ -41,7 +42,53 @@
 	GTEXT(_sys_resume_from_deep_sleep)
 #endif
 
+.macro install_page_tables
+#ifdef CONFIG_X86_MMU
+	/* Enable paging. If virtual memory is enabled, the instruction pointer
+	 * is currently at a physical address. There is an identity mapping
+	 * for all RAM, plus a virtual mapping of RAM starting at
+	 * CONFIG_KERNEL_VM_BASE using the same paging structures.
+	 *
+	 * Until we enable these page tables, only physical memory addresses
+	 * work.
+	 */
+	movl $Z_MEM_PHYS_ADDR(z_x86_kernel_ptables), %eax
+	movl %eax, %cr3
+
+#ifdef CONFIG_X86_PAE
+	/* Enable PAE */
+	movl %cr4, %eax
+	orl $CR4_PAE, %eax
+	movl %eax, %cr4
+
+	/* IA32_EFER NXE bit set */
+	movl $0xC0000080, %ecx
+	rdmsr
+	orl $0x800, %eax
+	wrmsr
+#endif /* CONFIG_X86_PAE */
+
+	/* Enable paging (CR0.PG, bit 31) / write protect (CR0.WP, bit 16) */
+	movl %cr0, %eax
+	orl $(CR0_PG | CR0_WP), %eax
+	movl %eax, %cr0
+
+	/* Jump to a virtual address, which works because the identity and
+	 * virtual mappings both are to the same physical address.
+	 */
+	lea vm_enter, %eax
+	jmp *%eax
+vm_enter:
```

Comment:
Sorry for the question, but I'm getting hard time to understand this construction, after you enabled paging loading the PD address in CR3 and setting the necessary bits in CR0, why you are loading the address of the label into eax to jump ? Also what it is the purpose ?

Reply:
the purpose is the comment a few lines above. the kernel boots always with the instruction pointer at the kernel's physical location in memory. if the kernel is linked at a virtual location (CONFIG_KERNEL_VM_BASE != CONFIG_SRAM_BASE_ADDRESS) this does the transition. it works because of the dual mapping set up in the page tables. see https://wiki.osdev.org/Higher_Half_x86_Bare_Bones

if CONFIG_KERNEL_VM_BASE == CONFIG_SRAM_BASE_ADDRESS (which is the current configuration) then this is a no-op. I should ifdef it.

Reply:
Thanks ! I think I got it. If CONFIG_KERNEL_BASE == CONFIG_SRAM_BASE_ADDRESS `vm_enter` == `&vm_enter`.

Reply:
yeah, the physical address and the virtual address are the same. I'm going to ifdef this to make a bit clearer.

Reply:
I added a `#if CONFIG_KERNEL_VM_BASE != CONFIG_SRAM_BASE_ADDRESS` block around this.